Improve Quantity Formatting Documentation

hl662 opened the pull request:
- Review existing documentation on Quantity Formatting found in core-frontend.
- Filled in documentation for core-quantity by introducing a new learnings section for quantity formatting outside of frontend use cases. Linked the documentation in core-quantity's README

ColinKerr commented on core/quantity/README.md:
```diff
@@ -2,10 +2,69 @@
+### Common Terms
+
+- [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.
+- [UnitsProvider]($quantity) - A class that will also locate the UnitProps for a unit given name or label. This class will also provide a [UnitConversion]($quantity) to convert from one unit to another.
```
you say a UnitsProvider 'will also locate' and 'will also provide' ... seems wrong for the first sentence to say 'will also' because nothing comes before it.

hl662 replied:
Fixed here and in the other QuantityFormatting.md

ColinKerr commented on core/quantity/README.md:
```diff
@@ -2,10 +2,69 @@
+
+- [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.
+- [UnitsProvider]($quantity) - A class that will also locate the UnitProps for a unit given name or label. This class will also provide a [UnitConversion]($quantity) to convert from one unit to another.
+- Unit Family/[Phenomenon]($ecschema-metadata) - The physical quantity that this unit measures (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.
```
what is 'this unit'?  Should this be written as 'A physical quantity that can be measured'

hl662 replied:
that actually was copy pasted from the other existing quantity formatting doc, I'll fix this here and in the other spot

hl662 replied:
Okay, fixed

ColinKerr commented on core/quantity/README.md:
```diff
@@ -2,10 +2,69 @@
+- [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.
+- [UnitsProvider]($quantity) - A class that will also locate the UnitProps for a unit given name or label. This class will also provide a [UnitConversion]($quantity) to convert from one unit to another.
+- Unit Family/[Phenomenon]($ecschema-metadata) - The physical quantity that this unit measures (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.
+- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel. While there is **NO** explicit flag or property to denote a persistence unit, those units are typically a part of the `Units.SI` [UnitSystem]($ecschema-metadata). The unit conversion of any persistence unit should both have its numerator and denominator set to 1.
```
This could be confusing: 'any persistence unit should both have it's numerator and denominator set to 1', I'm pretty sure that's true but so many other units have this true that I'm not sure it matters.  

Also this is mixing levels.  Persistence Unit is the unit used to persist the value, what we choose as the persistence unit is based on our schema standards.  There is nothing in core quantity that mandates that persistence unit be SI.

hl662 replied:
okay, so maybe I'll put the blurb about persistence unit in the learning doc in core-frontend?

hl662 replied:
QuantityFormatting.md already had a sentence on persistence unit from Bill, so I switched to using that

hl662 commented on core/quantity/README.md:
```diff
@@ -1,11 +1,230 @@
+
+### Common Terms
+
+- [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.
```
Instead of using $quantity here and everywhere else in this file, should I link straight to the line of code of the relevant interfaces and methods? Assuming this readme isn't part of the docs site

rschili replied:
I'm not familiar with using variables in markdown like this. Where do we define $quantity? The github preview for the file turns this into dead links.
The idea of putting line of code links here does not appeal to me. I've never seen that being done in abstract code documentation. People will read this .md to understand concepts. Also code links could easily go out of sync.

hl662 replied:
As an example, see this [README.md](https://github.com/iTwin/itwinjs-core/blob/master/docs/learning/frontend/QuantityFormatting.md?plain=1), and this [link](https://www.itwinjs.org/learning/frontend/quantityformatting/) shows how it is displayed in the docs site. So, these $<XXX_packagename> links really only make sense in the context of our itwinjs.org docs site. So I think I will remove the $quantity and everywhere else... I agree with the code links, I won't introduce them. Instead, I'll put a link at the top of the opening description, linking to the docs site's API reference for core-quantity

hl662 replied:
okay, added

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+  - [Contributing](#contributing)
+- [Licensing](#licensing)
+
+The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js references documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).
```
```suggestion
The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js reference documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js references documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).
+
+Also check the [iTwin.js learnings documentation](https://www.itwinjs.org/learning/frontend/quantityformatting/#quantity-package) explaining quantity formatting and it's basic concepts.
```
```suggestion
Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/frontend/quantityformatting/#quantity-package) explaining quantity formatting and its basic concepts.
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+### Common Terms
+
+- `Unit`/`UnitProps` - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.
+- `UnitsProvider` - An interface that locates the UnitProps for a unit given name or label. This interface also provides methods for `UnitConversion` to allow converting from one unit to another.
```
```suggestion
- `UnitsProvider` - An interface that locates the `UnitProps` for a unit given name or label. This interface also provides methods for `UnitConversion` to allow converting from one unit to another.
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel.
+- Format/FormatProp - The display format for the quantity value. For example, an angle may be persisted in radians but formatted and shown to user in degrees.
+  - CompositeValue - An addition to the format specification that allows the explicit specification of a unit label, it also allows the persisted value to be displayed as up to 4 sub-units. Typical multi-unit composites are used to display `feet'-inches"` and `degree°minutes'seconds"`.
+- `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the UnitsProvider during the formatting process.
```
```suggestion
- `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the `UnitsProvider` during the formatting process.
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+- Format/FormatProp - The display format for the quantity value. For example, an angle may be persisted in radians but formatted and shown to user in degrees.
+  - CompositeValue - An addition to the format specification that allows the explicit specification of a unit label, it also allows the persisted value to be displayed as up to 4 sub-units. Typical multi-unit composites are used to display `feet'-inches"` and `degree°minutes'seconds"`.
+- `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the UnitsProvider during the formatting process.
+- `ParserSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all other units in the same phenomenon. This is done to avoid async calls by the UnitsProvider and also done to allow a user to enter `43in` even when in "metric" unit system and have the string properly converted to meters.
```
```suggestion
- `ParserSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all other units in the same phenomenon. This is done to avoid async calls by the `UnitsProvider` and to allow users to input quantities in different unit systems than specified. For instance, if a metric unit system is set, a user could enter `43in` and have the result properly converted to meters. 
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+  - CompositeValue - An addition to the format specification that allows the explicit specification of a unit label, it also allows the persisted value to be displayed as up to 4 sub-units. Typical multi-unit composites are used to display `feet'-inches"` and `degree°minutes'seconds"`.
+- `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the UnitsProvider during the formatting process.
+- `ParserSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all other units in the same phenomenon. This is done to avoid async calls by the UnitsProvider and also done to allow a user to enter `43in` even when in "metric" unit system and have the string properly converted to meters.
+- `Formatter` - A class that holds methods to format a quantity value into a text string. Given a FormatterSpec object - containing one or many Unit definitions each with their own Unit Conversion info and a Format passed in - and a single `magnitude` number, the Formatter can convert that number into a text string, adhering to the properties passed to the formatTraits of a Format.
```
```suggestion
- `Formatter` - A class that holds methods to format a quantity value into a text string. Given a `FormatterSpec` object—which includes one or more unit definitions, each with their own conversion information and a specified format—and a single magnitude number, the `Formatter` can convert this number into a text string, adhering to the properties specified in the `formatTraits`.
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+- `FormatterSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all units defined in the format. This is done to avoid any async calls by the UnitsProvider during the formatting process.
+- `ParserSpec` - Holds the format specification as well as the `UnitConversion` between the persistence unit and all other units in the same phenomenon. This is done to avoid async calls by the UnitsProvider and also done to allow a user to enter `43in` even when in "metric" unit system and have the string properly converted to meters.
+- `Formatter` - A class that holds methods to format a quantity value into a text string. Given a FormatterSpec object - containing one or many Unit definitions each with their own Unit Conversion info and a Format passed in - and a single `magnitude` number, the Formatter can convert that number into a text string, adhering to the properties passed to the formatTraits of a Format.
+- `Parser` - A class that holds methods to parse a text string into a single number. Given a ParserSpec object containing a Format's Units and their Unit Conversions, as well as an input string, the Parser can either return an object `QuantityParseResult` that contains the magnitude of type `number`, or an object `ParseQuantityError`.
```
```suggestion
- `Parser` - A class that holds methods to parse a text string into a single number. Given a `ParserSpec` object containing a Format's Units and their Unit Conversions, as well as an input string, the Parser can either return an object `QuantityParseResult` that contains the magnitude of type `number`, or an object `ParseQuantityError`.
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+#### Units Provider
+
+To appropriately parse and output formatted values, A units provider is used to define all available units and provides conversion factors between units. There are a couple implementations of the UnitsProvider across iTwin.js:
```
```suggestion
To appropriately parse and output formatted values, a units provider is used to define all available units and provides conversion factors between units. There are several implementations of the UnitsProvider across iTwin.js:
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+The `BasicUnitsProvider` holds many common units and their conversions between each other.
+
+The `SchemaUnitProvider` is another example, used to load unit definitions of schemas from an iModel. This holds more extensive units through the Units schema, while also allowing users to define their own units.
```
```suggestion
The `SchemaUnitProvider` is used to load unit definitions of schemas from an iModel. This holds more extensive units through the Units schema, while also allowing users to define their own units.
```

hl662 replied:
addressed

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+#### Unit Conversion
+
+Unit conversion is performed through `UnitConversionSpec`. These objects are generated by a UnitsProvider, with the implementation determined by each specific provider. During initialization, a ParserSpec/FormatterSpec can ask for UnitConversionSpec objects provided via the UnitsProvider. During parsing and formatting, the specification will retrieve the UnitConversionSpec between the source and  ...
```
```suggestion
Unit conversion is performed through a `UnitConversionSpec`. These objects are generated by a `UnitsProvider`, with the implementation determined by each specific provider. During initialization, a `ParserSpec` or `FormatterSpec` can ask for `UnitConversionSpec` objects provided via the `UnitsProvider`. During parsing and formatting, the specification will retrieve the `UnitConversionSpec` between the source and destination units to apply the unit conversion.
```

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+Unit conversion is performed through `UnitConversionSpec`. These objects are generated by a UnitsProvider, with the implementation determined by each specific provider. During initialization, a ParserSpec/FormatterSpec can ask for UnitConversionSpec objects provided via the UnitsProvider. During parsing and formatting, the specification will retrieve the UnitConversionSpec between the source and  ...
+
+<!-- #### How a value is formatted
```
Stray comments? TODOs?

hl662 replied:
Yes, stray comments - I've removed them

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+#### Mathematical Operation Parsing
+
+The quantity formatter supports parsing mathematical operations. The operation is solved, formatting every values present, according to the specified format. This makes it possible to process several different units at once.
```
```suggestion
The quantity formatter supports parsing mathematical operations. The operation is solved, formatting each value present, according to the specified format. This makes it possible to process several different units at once.
```

hl662 replied:
addressed

ben-polinsky reviewed (commented):
nits/language

hl662 commented on core/quantity/README.md:
```diff
@@ -1,11 +1,232 @@
+
+### Common Terms
+
+- `Unit`/`UnitProps` - A named unit of measure which can be located by its name or label. The definition of any unit is represented through it's `UnitProps`.
```
addressed

hl662 commented:
@ColinKerr I've moved a majority of the stuff I wrote for core-quantity's README into the docs/learning section, updated the hyperlinks wherever needed. core-quantity's README on GH now points to this new learnings page. If you can just review the opening paragraph/section of this new index.md file, and the first heading that'd be great

aruniverse commented on docs/learning/quantity/index.md:
````diff
@@ -0,0 +1,215 @@
+<summary>Example Code</summary>
+
+```ts
+    const unitsProvider = new BasicUnitsProvider();
````
should prob add import lines

aruniverse reviewed (approved):
@ben-polinsky can you re-review as well?

ben-polinsky commented on core/quantity/README.md:
```diff
@@ -1,11 +1,15 @@
-Copyright © Bentley Systems, Incorporated. All rights reserved. See LICENSE.md for license terms and full copyright notice.
+The __@itwin/core-quantity__ package contains classes for quantity formatting and parsing. For detailed API documentation, see our [iTwin.js reference documentation](https://www.itwinjs.org/reference/core-quantity/quantity/).
+
+Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/quantity/index.md) explaining quantity formatting and its basic concepts.
```
Do you mean https://www.itwinjs.org/learning/frontend/quantityformatting ?

ben-polinsky replied:
```suggestion
Also check the [iTwin.js learning documentation](https://www.itwinjs.org/learning/quantity) explaining quantity formatting and its basic concepts.
```

ben-polinsky commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -42,11 +42,11 @@ To add custom labels use [QuantityFormatter.addAlternateLabels]($frontend) as sh
 ### Units Provider
 
-A units provider is used to define all available units and provides conversion factors between units. The [QuantityFormatter]($frontend) has a default units provider [BasicUnitsProvider]($frontend) that only defines units needed by the set of QuantityTypes the formatter supports. Most IModels contain a `Units` schema. If this is the case, an SchemaUnitsProvider may be defined when an IModel is op ...
+A units provider is used to define all available units and provides conversion factors between units. The [QuantityFormatter]($frontend) has a default units provider [BasicUnitsProvider]($frontend) that only defines units needed by the set of QuantityTypes the formatter supports. Most IModels contain a `Units` schema. If this is the case, a SchemaUnitProvider may be defined when an IModel is open ...
```
```suggestion
A units provider is used to define all available units and provides conversion factors between units. The [QuantityFormatter]($frontend) has a default units provider [BasicUnitsProvider]($frontend) that only defines units needed by the set of QuantityTypes the formatter supports. Most iModels contain a `Units` schema. If this is the case, a SchemaUnitProvider may be defined when an IModel is opened. The parent application must opt-in to using an iModel specific UnitsProvider using the following technique:
```

ben-polinsky commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -181,9 +181,10 @@ The Quantity Package `@itwinjs\core-quantity` defines interfaces and classes use
 Common Terms:
 
 - Unit/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label.
```
```suggestion
- Unit - A named unit of measure which can be located by its name or label. See [UnitProps]($quantity)
```

Is a bit better... but I don't love it.

ben-polinsky commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -181,9 +181,10 @@ The Quantity Package `@itwinjs\core-quantity` defines interfaces and classes use
-- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel.
+- [UnitsProvider]($quantity) - An interface that locates the UnitProps for a unit given name or label. This interface also provides methods for [UnitConversion]($quantity) to allow converting from one unit to another.
+- Unit Family/[Phenomenon]($ecschema-metadata) - A physical quantity that can be measured (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.
+- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel. IModels define the persistence unit through [KindOfQuantity](../../bis/ec/kindofquantity.md) objects.
```
```suggestion
- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable iModel. iModels define the persistence unit through [KindOfQuantity](../../bis/ec/kindofquantity.md) objects.
```

ben-polinsky replied:
I suppose we lose the links in gh with the $docs...

hl662 replied:
done

ben-polinsky commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -181,9 +181,10 @@ The Quantity Package `@itwinjs\core-quantity` defines interfaces and classes use
+- [UnitsProvider]($quantity) - An interface that locates the UnitProps for a unit given name or label. This interface also provides methods for [UnitConversion]($quantity) to allow converting from one unit to another.
+- Unit Family/[Phenomenon]($ecschema-metadata) - A physical quantity that can be measured (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.
+- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel. IModels define the persistence unit through [KindOfQuantity](../../bis/ec/kindofquantity.md) objects.
+- KindOfQuantity = An object that defines a persistence unit and presentation formats.
```
```suggestion
- [KindOfQuantity]($docs/bis/ec/kindofquantity/) - An object that defines a persistence unit and presentation formats.
```

hl662 replied:
done

ben-polinsky commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -181,9 +181,10 @@ The Quantity Package `@itwinjs\core-quantity` defines interfaces and classes use
+- Unit Family/[Phenomenon]($ecschema-metadata) - A physical quantity that can be measured (e.g., length, temperature, pressure).  Only units in the same phenomenon can be converted between.
+- Persistence Unit - The unit used to store the quantity value in memory or to persist the value in an editable IModel. IModels define the persistence unit through [KindOfQuantity](../../bis/ec/kindofquantity.md) objects.
+- KindOfQuantity = An object that defines a persistence unit and presentation formats.
 - Format/FormatProp - The display format for the quantity value. For example, an angle may be persisted in radians but formatted and shown to user in degrees.
```
I think it makes sense to link to terms (if the link exists) as they are defined even if they're linked elsewhere within the document?

hl662 replied:
yeah, especially in the common terms - I've updated that

ben-polinsky commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -195,13 +196,14 @@ For a detailed description of all the setting supported by FormatProp see the EC
 #### Numeric Format
 
-  The example below uses a simple numeric format and generates formatted string with 4 decimal place precision. For numeric formats there is no conversion to other units, the unit passed in is the unit returned with the unit label appended if "showUnitLabel" trait is set.
+  The example below uses a simple numeric format and generates a formatted string with 4 decimal place precision. For numeric formats there is no conversion to other units, the unit passed in is the unit returned with the unit label appended if "showUnitLabel" trait is set.
```
```suggestion
  The example below uses a simple numeric format and generates a formatted string with 4 decimal place precision. For numeric formats there is no conversion to other units; the unit passed in is the unit returned with the unit label appended if "showUnitLabel" trait is set.
```

hl662 replied:
done

ben-polinsky commented on docs/learning/quantity/index.md:
```diff
@@ -0,0 +1,225 @@
+
+### Common Terms
+
+- [Unit]($quantity)/[UnitProps]($quantity) - A named unit of measure which can be located by its name or label. The definition of any unit is represented through its `UnitProps`.
```
Do we need common terms here and in frontend/QuantityFormatting?

hl662 replied:
yeah now that these are both learning pages, it's redundant. So I've moved everything over to quantity's index page surrounding common terms, and added links to this page from QuantityFormatting.md

hl662 commented on docs/learning/frontend/QuantityFormatting.md:
```diff
@@ -195,13 +196,14 @@ For a detailed description of all the setting supported by FormatProp see the EC
 ### Formatting Examples
 
-Below are a couple examples of formatting values using methods directly from the @itwinjs/core-quantity package. The UnitsProvider used in the examples below can be seen [here](https://github.com/iTwin/itwinjs-core/blob/f3c8ee1d57ebb3e7b097a7e09201969fb57773ea/core/quantity/src/test/TestUtils/TestHelper.ts#L79). As discussed above, there are UnitProviders that can read units defined in the active ...
+Below are a couple examples of formatting values using methods directly from the @itwinjs/core-quantity package. The UnitsProvider used in the examples below can be seen in [TestHelper](https://github.com/iTwin/itwinjs-core/blob/master/core/quantity/src/test/TestUtils/TestHelper.ts). As discussed above, there are UnitProviders that can read units defined in the active IModel, and there is a basic ...
```
done
